sumpy/tools.py

```diff
@@ -125,6 +228,209 @@ def get_closest_cached_mi(self, mi):
+                self.cache_by_mi_q[(mi, q)] = expr
+                return expr
+
+        for i in range(self.dim):
+            if mi[i] >= 2:
+                mi_minus_one = list(mi)
+                mi_minus_one[i] -= 1
+                mi_minus_one = tuple(mi_minus_one)
+                mi_minus_two = list(mi)
+                mi_minus_two[i] -= 2
+                mi_minus_two = tuple(mi_minus_two)
+                expr = (mi[i]-1)*self.diff(mi_minus_two, q=q+1) * self.rscale ** 2
+                expr += self.var_list_multiplied[i] * self.diff(mi_minus_one, q=q+1)
+                expr = add_to_sac(self.sac, expr)
+                self.cache_by_mi_q[(mi, q)] = expr
+                return expr
+
+        assert mi == (0,)*self.dim
+        assert q > 0
+
+        prev_expr = self.diff(mi, q=q-1)
+        # Need to get expr.diff(r)/r, but we can only do expr.diff(x)
+        # Use expr.diff(x) = expr.diff(r) * x / r
+        expr = prev_expr.diff(self.var_list[0])/self.var_list[0]
+        # We need to distribute the division above
+        expr = expr.expand(deep=False)
+        self.cache_by_mi_q[(mi, q)] = expr
+        return expr
+
+
+class HelmholtzDerivativeTaker(RadialDerivativeTaker):
+    """Specialized derivative taker for Helmholtz potential.
+    """
+
+    def diff(self, mi, q=0):
+        import sumpy.symbolic as sym
+        if q < 2 or mi != (0,)*self.dim:
+            return RadialDerivativeTaker.diff(self, mi, q)
+        try:
+            return self.cache_by_mi_q[(mi, q)]
+        except KeyError:
+            pass
+
+        if self.dim == 2:
+            # See https://dlmf.nist.gov/10.6.E6
+            # and https://dlmf.nist.gov/10.6#E1
+            k = self.orig_expr.args[1] / self.r
+            expr = -  2 * (q - 1) * self.diff(mi, q - 1)
+            expr += - k**2 * self.diff(mi, q - 2)
+            expr /= self.r**2
+        else:
+            # See reference [Tausch2003] in RadialDerivativeTaker.diff
+            k = (self.orig_expr * self.r).args[-1] / sym.I / self.r
+            expr = -(2*q - 1)/self.r**2 * self.diff(mi, q - 1)
+            expr += -k**2 / self.r * self.diff(mi, q - 2)
+        self.cache_by_mi_q[(mi, q)] = expr
+        return expr
+
+
+@tag_dataclass
+class DifferentiatedExprDerivativeTaker:
+    """Implements the :class:`ExprDerivativeTaker` interface
+    for an expression that is itself a linear combination of
+    derivatives of a base expression. To take the actual derivatives,
+    it makes use of an underlying derivative taker *taker*.
+
+    .. attribute:: taker (ExprDerivativeTaker): A derivative taker for the
+            base kernel.
+    .. attribute:: derivative_transformation (dict): A dictionary mapping a
+            derivative multi-index to a coeffcient. The expression represented
+            by this derivative taker is the linear combination of the derivatives
+            of the expression for the base kernel.
```

Comment:
That doesn't render right:
![image](https://user-images.githubusercontent.com/352067/112382465-e09a9e80-8cb9-11eb-86f2-864f8e5e8d45.png)

I'll FTFY.